File: convex/training.ts
import { query, mutation } from "./_generated/server";
import { v } from "convex/values";
import { ConvexError } from "convex/values";
import { getCurrentUserOrThrow } from "./users.ts";

export const listCourses = query({
  args: { laboratoryId: v.id("laboratories") },
  handler: async (ctx, args) => {
    return await ctx.db.query("trainingCourses").withIndex("by_laboratory", (q) => q.eq("laboratoryId", args.laboratoryId)).collect();
  },
});

export const createCourse = mutation({
  args: {
    laboratoryId: v.id("laboratories"),
    title: v.string(),
    type: v.union(v.literal("sop"), v.literal("method"), v.literal("safety"), v.literal("instrument"), v.literal("regulatory"), v.literal("other")),
    description: v.optional(v.string()),
    linkedDocumentId: v.optional(v.id("documents")),
    linkedMethodId: v.optional(v.id("testMethods")),
    durationHours: v.optional(v.number()),
    validityMonths: v.optional(v.number()),
    assessmentRequired: v.boolean(),
    passingScore: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const user = await getCurrentUserOrThrow(ctx);
    const existing = await ctx.db.query("trainingCourses").withIndex("by_laboratory", (q) => q.eq("laboratoryId", args.laboratoryId)).collect();
    const courseCode = `TRN-${String(existing.length + 1).padStart(4, "0")}`;
    return await ctx.db.insert("trainingCourses", { ...args, courseCode, isActive: true, createdBy: user._id });
  },
});

export const listAssignments = query({
  args: { laboratoryId: v.id("laboratories"), userId: v.optional(v.id("users")) },
  handler: async (ctx, args) => {
    let assignments = await ctx.db.query("trainingAssignments").withIndex("by_laboratory", (q) => q.eq("laboratoryId", args.laboratoryId)).collect();
    if (args.userId) assignments = assignments.filter((a) => a.userId === args.userId);
    return await Promise.all(assignments.map(async (a) => {
      const course = await ctx.db.get(a.courseId);
      const user = await ctx.db.get(a.userId);
      return { ...a, courseTitle: course?.title, courseCode: course?.courseCode, courseType: course?.type, userName: user?.name };
    }));
  },
});

export const getMyAssignments = query({
  args: { laboratoryId: v.id("laboratories") },
  handler: async (ctx, args) => {
    const identity = await ctx.auth.getUserIdentity();
    if (!identity) return [];
    const user = await ctx.db.query("users").withIndex("by_token", (q) => q.eq("tokenIdentifier", identity.tokenIdentifier)).first();
    if (!user) return [];
    const assignments = await ctx.db.query("trainingAssignments").withIndex("by_user", (q) => q.eq("userId", user._id)).collect();
    return await Promise.all(assignments.map(async (a) => {
      const course = await ctx.db.get(a.courseId);
      return { ...a, courseTitle: course?.title, courseCode: course?.courseCode, courseType: course?.type };
    }));
  },
});

export const assignTraining = mutation({
  args: {
    courseId: v.id("trainingCourses"),
    userId: v.id("users"),
    laboratoryId: v.id("laboratories"),
    dueDate: v.optional(v.string()),
    notes: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    const user = await getCurrentUserOrThrow(ctx);
    const course = await ctx.db.get(args.courseId);
    if (!course) throw new ConvexError({ message: "Course not found", code: "NOT_FOUND" });
    return await ctx.db.insert("trainingAssignments", {
      ...args,
      assignedBy: user._id,
      assignedDate: new Date().toISOString(),
      status: "assigned",
    });
  },
});

export const completeTraining = mutation({
  args: {
    assignmentId: v.id("trainingAssignments"),
    completedDate: v.string(),
    assessmentScore: v.optional(v.number()),
    notes: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    await getCurrentUserOrThrow(ctx);
    const a = await ctx.db.get(args.assignmentId);
    if (!a) throw new ConvexError({ message: "Assignment not found", code: "NOT_FOUND" });
    const course = await ctx.db.get(a.courseId);
    let expiryDate: string | undefined;
    if (course?.validityMonths) {
      const expiry = new Date(args.completedDate);
      expiry.setMonth(expiry.getMonth() + course.validityMonths);
      expiryDate = expiry.toISOString();
    }
    await ctx.db.patch(args.assignmentId, { status: "completed", completedDate: args.completedDate, assessmentScore: args.assessmentScore, expiryDate, notes: args.notes });
  },
});

export const deleteAssignment = mutation({
  args: { assignmentId: v.id("trainingAssignments") },
  handler: async (ctx, args) => {
    await getCurrentUserOrThrow(ctx);
    await ctx.db.delete(args.assignmentId);
  },
});

export const getCompetencyMatrix = query({
  args: { laboratoryId: v.id("laboratories") },
  handler: async (ctx, args) => {
    const users = await ctx.db.query("users").collect();
    const courses = await ctx.db.query("trainingCourses").withIndex("by_laboratory", (q) => q.eq("laboratoryId", args.laboratoryId)).collect();
    const allAssignments = await ctx.db.query("trainingAssignments").withIndex("by_laboratory", (q) => q.eq("laboratoryId", args.laboratoryId)).collect();
    return { users: users.map((u) => ({ id: u._id, name: u.name ?? u.email ?? "Unknown" })), courses, assignments: allAssignments };
  },
});

File: backend/app/models/training.py
"""Mirrors convex/schema.ts `trainingCourses`, `trainingAssignments`
(see convex/training.ts)."""

import uuid

from sqlalchemy import Boolean, Float, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import LimsBase


class TrainingCourse(LimsBase):
    __tablename__ = "training_courses"

    laboratory_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("laboratories.id"), index=True)
    course_code: Mapped[str] = mapped_column(String(64), index=True)
    title: Mapped[str] = mapped_column(String(255))
    type: Mapped[str] = mapped_column(String(32))
    description: Mapped[str | None] = mapped_column(String(2000))
    linked_document_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("documents.id"))
    linked_method_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("test_methods.id"))
    duration_hours: Mapped[float | None] = mapped_column(Float)
    validity_months: Mapped[int | None] = mapped_column(Integer)
    assessment_required: Mapped[bool] = mapped_column(Boolean, default=False)
    passing_score: Mapped[float | None] = mapped_column(Float)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_by: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"))


class TrainingAssignment(LimsBase):
    __tablename__ = "training_assignments"

    course_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("training_courses.id"), index=True)
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"), index=True)
    laboratory_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("laboratories.id"), index=True)
    assigned_by: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"))
    assigned_date: Mapped[str] = mapped_column(String(32))
    due_date: Mapped[str | None] = mapped_column(String(32))
    completed_date: Mapped[str | None] = mapped_column(String(32))
    expiry_date: Mapped[str | None] = mapped_column(String(32))
    assessment_score: Mapped[float | None] = mapped_column(Float)
    status: Mapped[str] = mapped_column(String(16), default="assigned")
    notes: Mapped[str | None] = mapped_column(String(2000))
